src/packages/packages.controller.ts

```diff
@@ -0,0 +1,198 @@
+      if (current.minor > latest.minor) return current;
+      if (current.minor < latest.minor) return latest;
+      
+      if (current.patch > latest.patch) return current;
+      if (current.patch < latest.patch) return latest;
+      
+      return latest;
+    });
+    const download = new Download();
+    download.package = pkg;
+    download.version = latestVersion;
+    await this.em.persistAndFlush(download);
+  
+    const buffer = Buffer.from(latestVersion.packageBlob, 'base64');
+    const fileStream = new stream.PassThrough();
+    fileStream.end(buffer);
+  
+    const fileName = `${name}-${latestVersion.versionNumber}.tar.gz`;
+    res.set({
+      'Content-Type': 'application/octet-stream',
+      'Content-Disposition': `attachment; filename="${fileName}"`,
+      'Content-Length': buffer.length,
+    });
+  
+    fileStream.pipe(res);
+  }
+  
+  @Get(':name/:version/download')
+  async downloadPackage(
+    @Param('name') name: string,
+    @Param('version') version: string,
+    @Res() res: Response,
+  ) {
+    const [major, minor, patch] = version.split('.').map(Number);
+  
+    const pkg = await this.em.findOne(Package, { name }, { populate: ['versions'] });
+    if (!pkg) {
+      throw new NotFoundException(`Package "${name}" not found`);
+    }
+  
+    const ver = pkg.versions.getItems().find(
+      v => v.major === major && v.minor === minor && v.patch === patch,
+    );
+  
+    if (!ver) {
+      throw new NotFoundException(`Version "${version}" not found for package "${name}"`);
+    }
+  
+    const download = new Download();
+    download.package = pkg;
+    download.version = ver;
+    await this.em.persistAndFlush(download);
+  
+    const buffer = Buffer.from(ver.packageBlob, 'base64');
+    const fileStream = new stream.PassThrough();
+    fileStream.end(buffer);
+  
+    const fileName = `${name}-${version}.tar.gz`; 
+    res.set({
+      'Content-Type': 'application/octet-stream',
+      'Content-Disposition': `attachment; filename="${fileName}"`,
+      'Content-Length': buffer.length, 
+    });
+  
+    fileStream.pipe(res);
+  }
+  
+  @Get(':name/:version/downloads')
+  async getDownloadsHistory(
+    @Param('name') name: string,
+    @Param('version') version: string,
+  ) {
+    const pkg = await this.em.findOne(Package, { name });
+    if (!pkg) {
+      throw new NotFoundException(`Package "${name}" not found`);
+    }
+  
+    const [major, minor, patch] = version.split('.').map(Number);
+    
+    const ver = await this.em.findOne(Version, { 
+      package: pkg, 
+      major,
+      minor,
+      patch
+    });
+  
+    if (!ver) {
+      throw new NotFoundException(`Version "${version}" not found for package "${name}"`);
+    }
+  
+    const downloads = await this.em.find(Download, {
+      package: pkg,
+      version: ver
+    }, {
+      orderBy: { downloadDate: 'DESC' }
+    });
+  
+    return {
+      package: name,
+      version: version,
+      totalDownloads: downloads.length,
```

Comment:
if we have `downloads` - let's remove this property (it's just `downloads` array length)